src/blockchain_db/lmdb/db_lmdb.h

```diff
@@ -45,126 +46,215 @@ typedef struct txindex {
     tx_data_t data;
 } txindex;
 
-typedef struct mdb_txn_cursors
+enum class source : uint8_t
 {
-  MDB_cursor *m_txc_blocks;
-  MDB_cursor *m_txc_block_heights;
-  MDB_cursor *m_txc_block_info;
-
-  MDB_cursor *m_txc_output_txs;
-  MDB_cursor *m_txc_output_amounts;
-
-  MDB_cursor *m_txc_txs;
-  MDB_cursor *m_txc_txs_pruned;
-  MDB_cursor *m_txc_txs_prunable;
-  MDB_cursor *m_txc_txs_prunable_hash;
-  MDB_cursor *m_txc_txs_prunable_tip;
-  MDB_cursor *m_txc_tx_indices;
-  MDB_cursor *m_txc_tx_outputs;
-
-  MDB_cursor *m_txc_spent_keys;
-
-  MDB_cursor *m_txc_txpool_meta;
-  MDB_cursor *m_txc_txpool_blob;
-
-  MDB_cursor *m_txc_alt_blocks;
-
-  MDB_cursor *m_txc_hf_versions;
-
-  MDB_cursor *m_txc_properties;
-} mdb_txn_cursors;
-
-#define m_cur_blocks	m_cursors->m_txc_blocks
-#define m_cur_block_heights	m_cursors->m_txc_block_heights
-#define m_cur_block_info	m_cursors->m_txc_block_info
-#define m_cur_output_txs	m_cursors->m_txc_output_txs
-#define m_cur_output_amounts	m_cursors->m_txc_output_amounts
-#define m_cur_txs	m_cursors->m_txc_txs
-#define m_cur_txs_pruned	m_cursors->m_txc_txs_pruned
-#define m_cur_txs_prunable	m_cursors->m_txc_txs_prunable
-#define m_cur_txs_prunable_hash	m_cursors->m_txc_txs_prunable_hash
-#define m_cur_txs_prunable_tip	m_cursors->m_txc_txs_prunable_tip
-#define m_cur_tx_indices	m_cursors->m_txc_tx_indices
-#define m_cur_tx_outputs	m_cursors->m_txc_tx_outputs
-#define m_cur_spent_keys	m_cursors->m_txc_spent_keys
-#define m_cur_txpool_meta	m_cursors->m_txc_txpool_meta
-#define m_cur_txpool_blob	m_cursors->m_txc_txpool_blob
-#define m_cur_alt_blocks	m_cursors->m_txc_alt_blocks
-#define m_cur_hf_versions	m_cursors->m_txc_hf_versions
-#define m_cur_properties	m_cursors->m_txc_properties
-
-typedef struct mdb_rflags
+  blocks,
+  block_heights,
+  block_info,
+
+  output_txs,
+  output_amounts,
+
+  txs,
+  txs_pruned,
+  txs_prunable,
+  txs_prunable_hash,
+  txs_prunable_tip,
+  tx_indices,
+  tx_outputs,
+
+  spent_keys,
+
+  txpool_meta,
+  txpool_blob,
+
+  alt_blocks,
+
+  hf_versions,
+
+  properties
+};
+
+class mdb_txn_cursors
 {
-  bool m_rf_txn;
-  bool m_rf_blocks;
-  bool m_rf_block_heights;
-  bool m_rf_block_info;
-  bool m_rf_output_txs;
-  bool m_rf_output_amounts;
-  bool m_rf_txs;
-  bool m_rf_txs_pruned;
-  bool m_rf_txs_prunable;
-  bool m_rf_txs_prunable_hash;
-  bool m_rf_txs_prunable_tip;
-  bool m_rf_tx_indices;
-  bool m_rf_tx_outputs;
-  bool m_rf_spent_keys;
-  bool m_rf_txpool_meta;
-  bool m_rf_txpool_blob;
-  bool m_rf_alt_blocks;
-  bool m_rf_hf_versions;
-  bool m_rf_properties;
-} mdb_rflags;
-
-typedef struct mdb_threadinfo
+  public:
+    mdb_txn_cursors() noexcept :
+      m_cursors{{nullptr}}
+    {}
+
+    ~mdb_txn_cursors() noexcept
+    {
+      for (auto cursor: m_cursors) {
+        if (cursor != nullptr) {
+          mdb_cursor_close(cursor);
+        }
+      }
+    }
+
+    MDB_cursor*& get(source cursor)
+    {
+      return m_cursors.at(static_cast<uint8_t>(cursor));
```

Comment:
Can save some typing in the future by doing `uint8_t(cursor)`.

Reply:
I was unaware of this syntax. It does sound useful indeed. I'll remember this for when I need it.

Reply:
Was unfamiliar with that myself. Not sure it's worth using something that the rest of people don't recognize.